### ISSUE TITLE
Binary operators starting on new lines

### ISSUE THREAD

**Issue by edemaine** [DanielXMoore/Civet#1121]
I *think* this is a bug:
```coffee
extra.split(/\s+/).filter &
++ [state.start, state.goal]
```

Expected compilation:
```js
extra.split(/\s+/).filter($ => $)
.concat([state.start, state.goal])
```

Actual compilation:
```js
extra.split(/\s+/).filter($ =>
  $.concat([state.start, state.goal])
);
```

**Comment by STRd6** [comment]
Related #1280

**Comment by edemaine** [comment]
I notice that the same issue occurs without `&` altogether:

```js
extra.split(/\s+/).filter x
++ [state.start, state.goal]
```

The real issue is that we handle binary operators that start on the next line, and then there's a question of which expression to attach that to.

I guess our intuition is that the implicit function call (on `filter`) should end by the newline, but that's not how things work currently.

**Comment by edemaine** [comment]
Inspired by the bulleted arrays PR (#1361), I have an idea for how to address this: when entering `ImplicitArguments`, push an artificial indentation of the current indentation + 1, so that to remain inside the function call argument, you need to be strictly indented.  This would mean:

```js
filter foo
+ rest
↓↓↓
filter(foo)
+ rest

// but
filter foo
  + rest
↓↓↓
filter(foo
  + rest)
```

The bad news is that chaining multiple implicit function calls on the same line would require additional indentation, like so:

```js
f1 f2 f3 f4 f5 x
     + y
// ^needs 5 spaces because 5 function calls
```

I don't think that's good, but I'm also not quite sure how to fix it.

Maybe instead of increasing the indentation by 1, we add a boolean flag to the indentation saying "> i" where i is the current indentation, instead of the usual meaning of ">= i". And then we won't repeatedly add 1; instead we just repeatedly change the flag to ">" which has no effect beyond the first. Not the simplest solution, but maybe workable...

Let me know if you have better ideas!

**Comment by edemaine** [comment]
I reviewed this idea some more, and realize that we have *almost* this exact functionality, via `ForbidNewlineBinaryOp` (forbid newline followed by nested or indented binary op).  If I just add this to `ImplicitArguments`, then both

```js
extra.split(/\s+/).filter &
++ [state.start, state.goal]
```

```js
extra.split(/\s+/).filter &
  ++ [state.start, state.goal]
```

parse as

```js
extra
  .split(/\s+/)
  .filter(($) => $)
  .concat([state.start, state.goal]);
```

And no existing tests fail.

The question is whether we want this behavior, or whether we'd like the latter example to bring the `concat` inside the argument to `filter`. If the latter, it's tempting to redefine `ForbidNewlineBinaryOp` to mean forbid newline followed by nested binary op, but still allow binary ops that are indented further.  Unfortunately, we need to forbid the latter in examples like this:

```js
switch x
  > 0  // condition, not x > 0
```

So *if* we want these to parse differently, we could add `ForbidNestedBinaryOp`. Pretty similar to the `>` idea above, but just for binary operators. (And maybe it makes sense to use for other things too...)

---

I tried looking at how postfix accesses behave for comparison. [This one](http://civet.dev/playground?code=ZXh0cmEuc3BsaXQoL1xzKy8pLmZpbHRlciAmCi5zb3J0KCk%3D) makes sense:

```js
extra.split(/\s+/).filter &
.sort()
---
extra
  .split(/\s+/)
  .filter(($) => $)
  .sort();
```

But [this one](https://civet.dev/playground?code=ZXh0cmEuc3BsaXQoL1xzKy8pLmZpbHRlciAmCiAgLnNvcnQoKQ%3D%3D) does not:

```js
extra.split(/\s+/).filter &
  .sort()
---
($) => extra.split(/\s+/).filter & $.sort();
```

Presumably that's a bug, though, and `&` needs to be treated like a placeholder when it's followed by an access strictly indented on the next line instead of immediately after. Presumably we should match this behavior for binary operators too?
